modules/dbnd/test_dbnd/task/test_task_run_sync_local.py

```diff
@@ -0,0 +1,172 @@
+
+import pandas
+import pytest
+
+from dbnd import task
+from dbnd._core.parameter.parameter_builder import parameter
+from dbnd_test_scenarios.test_common.targets.target_test_base import TargetTestBase
+from targets import LocalFileSystem, target
+from targets.multi_target import MultiTarget
+from targets.target_config import TargetConfig
+
+
+logger = logging.getLogger(__name__)
+
+
+class PseudoLocalFileSystem(LocalFileSystem):
+    name = "pseudo_local"
+    local = False
+
+    def download(self, path, location, **kwargs):
+        shutil.copy(path, location)
+
+    def copy_from_local(self, local, dest, **kwargs):
+        shutil.copy(local, dest)
+
+
+class TestTaskRunSyncLocal(TargetTestBase):
+    @pytest.fixture
+    def my_target(self, pandas_data_frame):
+        _target = self.target(
+            "file.parquet",
+            fs=PseudoLocalFileSystem(),
+            config=TargetConfig().with_require_local_access(),
+        )
+        _target.write_df(pandas_data_frame)
+        return _target
+
+    @pytest.fixture
+    def my_second_target(self, pandas_data_frame):
+        _target = self.target(
+            "other_file.parquet",
+            fs=PseudoLocalFileSystem(),
+            config=TargetConfig().with_require_local_access(),
+        )
+        _target.write_df(pandas_data_frame)
+        return _target
+
+    @pytest.fixture
+    def my_multitarget(self, my_target, my_second_target):
+        return MultiTarget([my_target, my_second_target])
+
+    @pytest.fixture
+    def test_task(self, my_target):
+        @task
+        def t_f(input_=parameter[pandas.DataFrame]):
+            return input_
+
+        return t_f
+
+    @pytest.fixture
+    def create_local_multitarget(self, my_multitarget):
+        def _local_multitarget(cache_folder):
+            return MultiTarget(
+                [
+                    target(
+                        cache_folder,
+                        os.path.basename(subtarget.path),
+                        config=subtarget.config,
+                    )
+                    for subtarget in my_multitarget.targets
+                ]
+            )
+
+        return _local_multitarget
+
+    def test_task_run_sync_sync_multi_target(
+        self,
+        test_task,
+        my_target,
+        my_second_target,
+        my_multitarget,
+        create_local_multitarget,
+    ):
+        test_task = test_task.t(my_multitarget)
+        task_run = test_task.dbnd_run().root_task_run
+        sync_local = task_run.sync_local
+
+        assert len(sync_local.inputs_to_sync) == 1
+        task_param, old_multitarget, new_multitarget = sync_local.inputs_to_sync[0]
+
+        assert task_param == test_task._params.get_param("input_")
+        assert old_multitarget == my_multitarget
+
+        local_multitarget = create_local_multitarget(task_run.attemp_folder_local_cache)
+
+        self.compare_multitargets(new_multitarget, local_multitarget)
+        with patch.object(
```

Comment:
Dupped code with the test below, merge them :)

Reply:
While I agree that code in both tests is similar I'm not sure if merging both tests into one is a good idea - such code would difficult to read with a lot of code nested inside if/else statements (because the testing flow is a bit different for `MultiTarget` and for `FileTarget`). What do you think?

Reply:
Definitely export the mock patch ctx manager, other parts are less critical

Reply:
fixd